### DISCUSSION TITLE
NavigationStack(path:root:destination) Not Working

### DISCUSSION THREAD

**Discussion by jondlew** [pointfreeco/swift-composable-architecture#2756] [General]
### Description

`NavigationStack(path:root:destination)` has stopped working on iOS 16.

Navigation using this method still works under observation-beta, but has stopped working on main.

I've linked to a sample project exhibiting the issue. Additionally, the behavior on iOS 17 seems strange, as the state.path doesn't seem to be popped when using the back navigation button.

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

Tap on a list item to navigate to detail

### Actual behavior

No navigation occurs on iOS 16. 

On iOS 17, navigation works, but the navigation path is not popped when using the back button or swipe to go back. Tapping on multiple list items causes the stack to grow, and you have to navigate back multiple times to reach the list view.

### Steps to reproduce

Here is a project that exhibits the issue: [NavigationStackTester](https://github.com/jondlew/NavigationStackTester)

### The Composable Architecture version information

main

### Destination operating system

iOS 16 and iOS 17

### Xcode version information

15.2

**Comment by mbrandonw** [comment]
Hi @jondlew, I'm not sure I understand the issue. `NavigationStack` is only available on iOS 16+, so how are you running it on iOS 14/15? Even the project you provided is limited to iOS 16.1+.

**Comment by jondlew** [comment]
Sorry -- brainfart. Replace all iOS 14/15 with iOS 16/17!

So NavigationStack is working on iOS 17, but stopped working on iOS 16.

Clearly still living in the past -- thanks!

**Comment by mbrandonw** [comment]
It looks like you forgot to use `forEach` in the parent reducer:

```swift
public var body: some ReducerOf<Self> {
  Reduce { state, action in
    // ...
  }
  .forEach(\.path, action: \.path) {
    Path()
  }
}
```

That should fix things. Since this isn't a problem with the library I am going to convert it to a discussion.

**Comment by jondlew** [comment]
Yes, I just noticed this, and it does fix the iOS 17 issue with the back button.

The navigation is still broken in iOS 16, however. Can you please double check this under iOS 16?

Thanks!

**Reply by mbrandonw** [reply]
Yeah, there is something strange happening in iOS 16. We've tracked it down to be the fault of `@State`. If you use `@Perception.Bindable` instead it will work fine.

We have found that bindings derived from `@State` behave differently than bindings derived from other things (e.g. `@Bindable`, `@ObservedObject`, etc.). And for some reason the changes in #2740 do not play nicely with bindings derived from `@State`. 

The fact that it works in iOS 17 makes us think it was likely a bug in `@State` that is now fixed. So we are going to keep the existing behavior, and we just think there are not a lot of reasons to use `@State`. You should just use either `let` or `@Bindable var`.

**Reply by jondlew** [reply]
Okay, sounds good. @Perception.Bindable works fine -- Thanks!
